relay-general/src/store/normalize/spans.rs

```diff
@@ -0,0 +1,727 @@
+            // current interval as the reference going forwards
+            non_overlapping_intervals.push(previous);
+            previous = current.clone();
+        }
+    }
+
+    // Make sure to include the final interval.
+    non_overlapping_intervals.push(previous);
+
+    non_overlapping_intervals
+}
+
+/// Computes the exclusive time of the source interval after subtracting the
+/// list of intervals.
+/// Assumes that the input intervals are sorted by start time.
+fn interval_exclusive_time(source: &TimeInterval, intervals: &[TimeInterval]) -> f64 {
+    let mut exclusive_time = 0.0;
+
+    let mut remaining = source.clone();
+
+    for interval in intervals {
+        if interval.end < remaining.start {
+            // The interval is entirely to the left of the remaining interval,
+            // so nothing to be done here.
+            continue;
+        } else if interval.start >= remaining.end {
+            // The interval is entirely to the right of the remaining interval,
+            // so nothing to be done here.
+            //
+            // Additionally, since intervals are sorted by start time, all
+            // intervals afterwards can be skipped.
+            break;
+        } else {
+            // The interval must intersect with the remaining interval in some way.
+
+            if interval.start > remaining.start {
+                // The interval begins within the remaining interval, there is a
+                // portion to its left that should be added to the results.
+                exclusive_time += TimeInterval::new(remaining.start, interval.start).get_duration();
+            }
+
+            if interval.end < remaining.end {
+                // The interval ends within the remaining interval, so the
+                // tail of the interval interesects with the head of the remaining
+                // interval.
+                //
+                // Subtract the intersection by shifting the start of the remaining
+                // interval.
+                remaining.start = interval.end;
+            } else {
+                // The interval ends to the right of the remaining interval, so
+                // the interval intersects with the entirety of the remaining
+                // interval. So zero out the interval.
+                remaining.start = remaining.end;
+
+                // There is nothing remaining to be checked.
+                break;
+            }
+        }
+    }
+
+    // make sure to add the remaining interval
+    exclusive_time + remaining.get_duration()
+}
+
+fn get_span_interval(span: &Span) -> Option<TimeInterval> {
+    let start_timestamp = *span.start_timestamp.value()?;
+    let end_timestamp = *span.timestamp.value()?;
+    Some(TimeInterval::new(start_timestamp, end_timestamp))
+}
+
+pub fn normalize_spans(event: &mut Event, attributes: &BTreeSet<SpanAttribute>) {
+    if attributes.contains(&SpanAttribute::ExclusiveTime) {
```

Comment:
_nit_: Make this a `match` statement. That way, once a new attribute is added, the compiler will point out this place for needing an implementation. You could also move the actual code into a sub-function.

Reply:
I'm not sure what you mean by the compiler will point out this will need additional implementation. This is a `attributes.contains` check so a `match` will only require an implementation for `true` and false` right?

Reply:
Ah sorry, this wasn't well explained. I was thinking along these lines:

```rust
for attribute in attributes {
    match attribute {
        SpanAttribute::ExclusiveTime => { ... }
        SpanAttribute::Unknown => (), // ignored
    }
}
```